### ISSUE TITLE
Customizable note headings

### ISSUE THREAD

**Issue by dvanoni** [dvanoni/notero#316]
![CleanShot 2023-08-11 at 20 21 36@2x](https://github.com/dvanoni/notero/assets/299357/6c6eaaaf-f497-4d67-a5f9-dccaf7609dc5)

To make it a little complicated: I would love A to be dropdown heading 2 and B to be dropdown heading 3. This fits my notion page the best. Finally the text in B is also referenced again below. Would it be possible to remove that part?

_Originally posted by @ThomasDeLange in https://github.com/dvanoni/notero/issues/4#issuecomment-1676904491_

**Comment by dvanoni** [comment]
@ThomasDeLange, I pulled this out into a new issue so that we can discuss it separately.

It should be possible to allow customization of the heading levels used for the synced notes. I can add these as options in the Notero preferences, but this will be lower priority than some other improvements I'd like to get in place first.

---

As for the duplication of heading **B**, this is a bit tricky. What's happening here is that the _title_ of the note item is used as the toggle heading (**B**), and the _full content_ of the note is a child of the toggle heading.

![CleanShot 2023-08-14 at 19 06 46@2x](https://github.com/dvanoni/notero/assets/299357/80f80468-d8b4-4038-93ce-b68cae921b0f)

As far as I can tell, Zotero uses the first line of the note content as the title of the note item. When the note has a heading at the start of it, you end up with the duplicated headings in Notion as you've seen.

The approach I've currently implemented is definitely the simplest since I don't have to do any special logic—I just use the title for the title and the content for the content. I could try to implement an approach where I omit the title from the content, but because of how the content is structured, this could be a little tricky. Another possible downside is that we could lose any formatting from the first line of a note (as in the screenshot above).

I agree that the current implementation doesn't look super great when the note starts with a heading, but I'm also struggling to come up with the best alternative that would support all scenarios. I'm open to suggestions!

**Comment by ThomasDeLange** [comment]
@dvanoni I understand, maybe some basic pattern matching or regex could help remove the title from the text? Since they are exactly the same this should work. Otherwise maybe counting the length of the title and removing that from the front of the text.

**Comment by dvanoni** [comment]
Yeah, I think that would work. For scenarios where the first line of the note is a plain heading—like with the extracted annotations—it should be straightforward to match against the title text. And I think it's those scenarios where it's most desirable to have that duplicate heading omitted from Notion. 👍

**Comment by dvanoni** [comment]
From @Pierre-Debraux in https://github.com/dvanoni/notero/issues/4#issuecomment-1853313031:

> Would it be possible to not have a few levels of indentation ?
>
> <img alt="image" width="918" src="https://private-user-images.githubusercontent.com/151032182/290089281-3972f647-ff5c-4154-afc4-188cfc0fd7af.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI0NDcyMzksIm5iZiI6MTcwMjQ0NjkzOSwicGF0aCI6Ii8xNTEwMzIxODIvMjkwMDg5MjgxLTM5NzJmNjQ3LWZmNWMtNDE1NC1hZmM0LTE4OGNmYzBmZDdhZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxM1QwNTU1MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZjVhZWJiZTQyNWE0NzNkMzFmYjYzYjFjNjQyNTg1MTI3Zjk5ZjJkMTAwODMzMjY2MWY4YWRlZTk4MjQ3Y2M0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.a6-DU0EMGAQEE9-5kvwHUknQwy_zs5NZEa317Eug_xY">

I agree the indentation is not ideal. It's currently implemented this way for technical reasons, specifically:

- The top-level "Zotero Notes" toggle heading provides a container where Notero knows it can safely manage the content. By staying within this confined container, it avoids conflicting with anything else that a user may have entered into the Notion page themselves. I can think of some options around this:
   - Allow the user to disable the heading altogether. Notero would place notes directly on the page—which should be fine as long as the user doesn't intend to modify the page themselves.
   - Keep the heading, but don't make it a toggle heading (i.e. remove the indentation). As of [earlier this year](https://developers.notion.com/page/changelog#june-13---july-10-2023), the Notion API now supports an `after` parameter which would allow us to place content directly after the "Zotero Notes" heading without having to be indented.
- The individual note toggle headings also provide a container for Notero to place their content. When a note is updated in Zotero, we actually delete the entire toggle heading and its children (the note content) and then recreate the note under a new toggle heading. We use this approach because of how the Notion API is structured. If we didn't have the note's content within a single container, updating a note would require deleting every piece of its content one by one. I've made an assumption that this would be too inefficient—especially for large notes—but it could be worth exploring to see how it behaves.

So, there are some options we may be able to try. 🙂 I'll share progress here when I'm able to dig into this.

**Comment by Pierre-Debraux** [comment]
@dvanoni from my perspective, I think the top-level "Zotero notes" is good as it is. And as you explain, that way, Zotero knows where to put the notes and in the mean time, we can add whatever we want on the same page, but the notes are always at the same place.

It's more the "Annotations" toggle and then, the "Annotations" title that "annoy" me the most.

Still, it's no big deal, of course. But I think, they're unnecessary, at least in my use case.
